bd-report-writer/src/ffi.rs

```diff
@@ -0,0 +1,569 @@
+/// containing the report
+///
+/// # Safety
+/// This function returns a null pointer in the event that the `handle`
+/// pointer is invalid. It should also only be called **once**.
+#[unsafe(no_mangle)]
+extern "C-unwind" fn bdrw_get_completed_buffer(
+  handle: BDProcessorHandle,
+  buffer_length: *mut u64,
+) -> *const u8 {
+  let mut processor = unpack_processor!(handle, null());
+  let binary_images = processor
+    .builder
+    .create_vector(processor.binary_images.as_slice());
+  let threads = processor
+    .builder
+    .create_vector(processor.threads.as_slice());
+  let thread_details = if processor.system_thread_count > 0 || !processor.threads.is_empty() {
+    Some(ThreadDetails::create(
+      &mut processor.builder,
+      &ThreadDetailsArgs {
+        count: processor.system_thread_count,
+        threads: Some(threads),
+      },
+    ))
+  } else {
+    None
+  };
+  let errors = processor.builder.create_vector(processor.errors.as_slice());
+  let report = Report::create(
+    &mut processor.builder,
+    &ReportArgs {
+      binary_images: Some(binary_images),
+      sdk: Some(processor.sdk),
+      type_: processor.report_type,
+      app_metrics: processor.app,
+      device_metrics: processor.device,
+      errors: Some(errors),
+      thread_details,
+    },
+  );
+  processor.builder.finish(report, None);
+  let data = processor.builder.finished_data();
+  if !buffer_length.is_null() {
+    unsafe {
+      *buffer_length = data.len() as u64;
+    }
+  }
+  let ptr = data.as_ptr();
+  pack_processor!(handle, processor);
+  ptr
+}
+
+/// Disposes of the handle and *the associated memory buffer*.
+///
+/// # Safety
+/// All actions involving the memory buffer returned from
+/// `bdrw_get_completed_buffer()` should be completed prior to calling this
+/// function.
+#[unsafe(no_mangle)]
+extern "C" fn bdrw_dispose_buffer_handle(handle: BDProcessorHandle) {
+  unsafe {
+    drop(Box::<ReportProcessor<'_>>::from_raw(*handle as *mut _));
+    *handle = null();
+  }
+}
+
+/// Add a new binary image to the report. A valid value for both `image.id`
+/// and `image.path` is required to add the image to the report or else the
+/// input is discarded (returning `false`).
+#[unsafe(no_mangle)]
+extern "C-unwind" fn bdrw_add_binary_image(
+  handle: BDProcessorHandle,
+  image: *const BDBinaryImage,
+) -> bool {
+  let mut processor = unpack_processor!(handle, false);
+  if let Some(unpacked_image) = unsafe { image.as_ref() }.and_then(|image| {
+    let id = append_string(&mut processor.builder, image.id);
+    let path = append_string(&mut processor.builder, image.path);
+    if id.is_some() && path.is_some() {
+      Some(BinaryImage::create(
+        &mut processor.builder,
+        &BinaryImageArgs {
+          id,
+          path,
+          load_address: image.load_address,
+        },
+      ))
+    } else {
+      None
+    }
+  }) {
+    processor.binary_images.push(unpacked_image);
+    pack_processor!(handle, processor);
```

Comment:
instead of pack/unpack you could just get a &mut handle to the processor, you're already assuming exclusive access so it could clean it up a little bit